src/couch_views/src/couch_views_reader.erl

```diff
@@ -0,0 +1,204 @@
+% Licensed under the Apache License, Version 2.0 (the "License"); you may not
+% use this file except in compliance with the License. You may obtain a copy of
+% the License at
+%
+%   http://www.apache.org/licenses/LICENSE-2.0
+%
+% Unless required by applicable law or agreed to in writing, software
+% distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+% WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+% License for the specific language governing permissions and limitations under
+% the License.
+
+-module(couch_views_reader).
+
+-export([
+    read/6
+]).
+
+
+-include("couch_views.hrl").
+-include_lib("couch/include/couch_db.hrl").
+-include_lib("fabric/src/fabric2.hrl").
+
+
+read(Db, DDoc, ViewName, Callback, Acc0, Args) ->
+    #{name := DbName} = Db,
+
+    {ok, Mrst} = couch_views_util:ddoc_to_mrst(DbName, DDoc),
+    #mrst{
+        sig = Sig,
+        views = Views
+    } = Mrst,
+
+    IdxName = get_idx_name(ViewName, Views),
+    State0 = #{
+        acc => Acc0,
+        skip => maps:get(skip, Args, 0),
+        include_docs => maps:get(include_docs, Args, false),
+        db => Db
+    },
+
+    DefaultOpts = [{streaming_mode, want_all}],
+    {Start, End, QueryOpts} = convert_args_to_fdb(Db, Sig, IdxName, Args,
+        DefaultOpts),
+    Opts = QueryOpts ++ DefaultOpts,
+
+    fabric2_fdb:transactional(Db, fun(TxDb) ->
+        Future = couch_views_fdb:get_map_range(TxDb, Start, End, Opts),
+
+        UnPack = get_unpack_fun(TxDb, Opts, Callback),
+        State1 = lists:foldl(UnPack, State0, erlfdb:wait(Future)),
+
+        #{acc := Acc1} = State1,
+        Callback(complete, Acc1)
+    end).
+
+
+get_idx_name(ViewName, Views) ->
+    {value, View} = lists:search(fun (View) ->
+        lists:member(ViewName, View#mrview.map_names)
+    end, Views),
+    View#mrview.id_num.
+
+
+convert_args_to_fdb(Db, Sig, IdxName, Args, Opts) ->
+    #{
+        direction := Direction
+    } = Args,
+
+    {Start1, End1} = get_range_keys(Db, Sig, IdxName, Args),
+
+    Opts1 = case maps:is_key(limit, Args) of
+        false ->
+            Opts;
+        true ->
+            Skip = maps:get(skip, Args, 0),
+            Limit = maps:get(limit, Args),
```

Comment:
Do we need a default value there as well, or we are guaranteed to have one?

Reply:
we set one in `couch_views.erl`